KAFKA-13933: Fix stuck SSL tests in case of authentication failure

fvaleri opened the pull request:
When there is an authentication error after the initial TCP connection,
the selector never becomes READY, and these tests wait forever waiting for this state.

This is actually what happened to me while using an OpenJDK build that does not
support the required cipher suites.

fvaleri committed:
KAFKA-13933: Fix stuck SSL unit tests in case of authentication failure

When there is an authentication error after the initial TCP connection,
the selector never becomes READY, and these tests wait forever waiting for this state.

This is actually what happened to me while using an OpenJDK build that does not
support the required cipher suites.

Signed-off-by: Federico Valeri <fedevaleri@gmail.com>

divijvaidya commented on clients/src/test/java/org/apache/kafka/common/network/SelectorTest.java:
```diff
@@ -449,9 +450,9 @@ public void close() {
-
-        time.sleep(6000); // The max idle time is 5000ms
+        selector.connect(id, new InetSocketAddress("localhost", server.port), BUFFER_SIZE, BUFFER_SIZE);
+        selector.poll(0);
```
Note that `connect()` is synchronous method here. It might not complete with the first poll event. That is the reason it is tested in a while loop inside `blockingConnect`.

My suggestion would be to keep using blockingConnect with a (new) timeout to terminate while loop in it similar to implementation of `NetworkTestUtils.waitForChannelReady`.

fvaleri replied:
Hi @divijvaidya, thanks. 

I need to fix this and `testCloseOldestConnectionWithMultiplePendingReceives`.

fvaleri commented:
@divijvaidya I fixed the test you were referring to.

I also fixed the help method `sendNoReceive` which was directly using `channel.mute()` instead of `selector.mute(channel.id())`. The first call does not ensure proper state handling, and I believe it's the reason why `testCloseOldestConnectionWithMultiplePendingReceives` was randomly failing.

Let's see if the test job pass this time.

fvaleri commented:
All selector's tests are passing. Failing tests seem to be unrelated.

showuon commented on clients/src/test/java/org/apache/kafka/common/network/SelectorTest.java:
```diff
@@ -603,23 +603,18 @@ private void verifyChannelExpiry(KafkaChannel channel) throws Exception {
+            time.sleep(CONNECTION_MAX_IDLE_MS + 1_000);
+            selector.poll(completedReceives == expectedReceives ? 0 : 1_000);
             completedReceives += selector.completedReceives().size();
-            if (!selector.completedReceives().isEmpty()) {
-                assertEquals(1, selector.completedReceives().size());
-                assertNotNull(selector.channel(id), "Channel should not have been expired");
-                assertTrue(selector.closingChannel(id) != null || selector.channel(id) != null, "Channel not found");
-                assertFalse(selector.disconnected().containsKey(id), "Disconnect notified too early");
-            }
```
Why did we remove these assetions?

fvaleri replied:
They are redundant IMO. 

If any of these fails, you would catch it in the final assertions.

showuon replied:
You're right! Make sense!

ijuma replied:
Can you elaborate why this is so? The assertions here are different from the ones after the while loop.

fvaleri replied:
Hi @ijuma, thanks for looking into this.

You are right, I think the only missing assertion is the following.

```
assertFalse(selector.disconnected().containsKey(id), "Disconnect notified too early");
```

If disconnected() is not empty, then we would exit the while loop prematurely and the first assertion would fail. Maybe we could simply add a message which also includes the channel's state to the first assertion after the while loop.

showuon commented on clients/src/test/java/org/apache/kafka/common/network/SelectorTest.java:
```diff
@@ -1138,20 +1130,18 @@ private KafkaChannel createConnectionWithPendingReceives(int pendingReceives) th
+    private void sendNoReceive(KafkaChannel channel, int numRequests) throws Exception {
+        selector.mute(channel.id());
         for (int i = 0; i < numRequests; i++) {
             selector.send(createSend(channel.id(), String.valueOf(i)));
             do {
                 selector.poll(10);
             } while (selector.completedSends().isEmpty());
         }
-        channel.maybeUnmute();
-
-        return channel;
+        selector.unmute(channel.id());
```
nice clean up.

fvaleri replied:
This is not just a cleanup, but mainly a bug fix. 

Using channel's mute/unmute rather than selector's mute/unmute does not ensure proper selector state handling, which causes random errors when a test rely on that state.

showuon commented on clients/src/test/java/org/apache/kafka/common/network/SelectorTest.java:
```diff
@@ -1030,10 +1025,7 @@ private void blockingConnect(String node) throws IOException {
-            selector.poll(10000L);
-        while (!selector.isChannelReady(node))
-            selector.poll(10000L);
+        NetworkTestUtils.waitForChannelReady(selector, node);
```
I think we should replace the 2 while loop with:
`waitForChannelConnected` + `waitForChannelReady`.
This way, we can make sure the connection is done or not when exception thrown during `waitForChannelReady`, right?

Same comments to below similar replacement.

fvaleri replied:
Let me add the KafkaChannel's state machine diagram for clarity. 

Given that the TCP connection is already established after connect completes successfully (state: READY|AUTHENTICATE), I don't think it is necessary to wait for channel readiness, which would unnecessary delay this test in case of TLS/SASL authentications. Instead, it is necessary for all the other tests on this class.

![kafka-channel-states](https://user-images.githubusercontent.com/11456498/169981699-f0e7fc6c-985b-407a-80c7-8294643f40f8.jpg)

tombentley replied:
The question is whether this test should assume the correctness of the implementation of that state machine in KafkaChannel. It's probably a safe assumption, but I'd be inclined to not make that assumption.

fvaleri replied:
@showuon I probably expressed that badly. I added `NetworkTestUtils.waitForChannelConnected()` for tests that don't need to use the channel, like `SelectorTest.testCloseOldestConnection()`, where we test for idle connection expiration, so there is no need to wait for readiness.

@tombentley agreed.

showuon commented on clients/src/test/java/org/apache/kafka/common/network/NetworkTestUtils.java:
```diff
@@ -87,6 +88,16 @@ public static void checkClientConnection(Selector selector, String node, int min
+                && !selector.channel(node).isConnected() && secondsLeft-- > 0) {
+            selector.poll(1000L);
+        }
+        assertNotNull(selector.channel(node));
+        assertTrue(selector.channel(node).isConnected());
```
I'll add error message in the assertion, to indicate what we expected, ex:
`assertTrue(selector.channel(node).isConnected(), "channel " + node + " is not connected after 30 secs");`

Same to below `waitForChannelReady` method. (I know it's not your change, but let's improve it together)

fvaleri replied:
Agreed, thanks.

showuon reviewed (commented):
Thanks for the improvement! Yes, it's not good to have a long waiting block in the codes. Left some comments. And also, the PR title should start with `MINOR`. Ex: `MINOR: Fix stuck SSL tests in case of authentication failure`. Thanks.

tombentley commented on clients/src/test/java/org/apache/kafka/common/network/NetworkTestUtils.java:
```diff
@@ -87,13 +88,23 @@ public static void checkClientConnection(Selector selector, String node, int min
+            selector.poll(1000L);
+        }
+        assertNotNull(selector.channel(node));
+        assertTrue(selector.channel(node).isConnected(), String.format("Channel %s is not connected after %d seconds", node, secondsLeft));
```
When the assertion fails `secondsLeft` will be zero at this point (since you decremented it in the loop) resulting in a confusing assertion failure message.

fvaleri replied:
Right, thanks for spotting this.

showuon reviewed (approved):
LGTM! Thanks for the fix!

showuon commented:
Failed tests are unrelated:
```
Build / JDK 17 and Scala 2.13 / kafka.admin.TopicCommandIntegrationTest.testDescribeAtMinIsrPartitions(String).quorum=kraft
    Build / JDK 17 and Scala 2.13 / kafka.controller.ControllerIntegrationTest.testPartitionReassignmentToBrokerWithOfflineLogDir()
    Build / JDK 17 and Scala 2.13 / org.apache.kafka.streams.integration.NamedTopologyIntegrationTest.shouldAllowRemovingAndAddingNamedTopologyToRunningApplicationWithMultipleNodesAndResetsOffsets
```

showuon commented:
@tombentley , do you want to have another look? I'm going to merge this PR next week if you don't have any other comments. Thank you.
